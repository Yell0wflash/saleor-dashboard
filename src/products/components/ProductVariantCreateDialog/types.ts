export type ProductVariantCreateStep = "values" | "prices" | "summary";
