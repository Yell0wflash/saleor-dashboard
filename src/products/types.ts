import { FilterOpts, MinMax } from "@saleor/types";
import { StockAvailability } from "@saleor/types/globalTypes";

export enum ProductStatus {
  PUBLISHED = "published",
  HIDDEN = "hidden"
}

export interface ProductListFilterOpts {
  price: FilterOpts<MinMax>;
  status: FilterOpts<ProductStatus>;
  stockStatus: FilterOpts<StockAvailability>;
}
