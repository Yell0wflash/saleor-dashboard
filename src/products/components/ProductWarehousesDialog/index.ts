export { default } from "./ProductWarehousesDialog";
export * from "./ProductWarehousesDialog";
