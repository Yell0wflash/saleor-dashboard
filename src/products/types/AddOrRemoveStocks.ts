/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

import { StockInput, ProductErrorCode, StockErrorCode } from "./../../types/globalTypes";

// ====================================================
// GraphQL mutation operation: AddOrRemoveStocks
// ====================================================

export interface AddOrRemoveStocks_productVariantStocksCreate_errors {
  __typename: "BulkStockError";
  code: ProductErrorCode;
  field: string | null;
  index: number | null;
}

export interface AddOrRemoveStocks_productVariantStocksCreate_productVariant_stocks_warehouse {
  __typename: "Warehouse";
  id: string;
  name: string;
}

export interface AddOrRemoveStocks_productVariantStocksCreate_productVariant_stocks {
  __typename: "Stock";
  id: string;
  quantity: number;
  quantityAllocated: number;
  warehouse: AddOrRemoveStocks_productVariantStocksCreate_productVariant_stocks_warehouse;
}

export interface AddOrRemoveStocks_productVariantStocksCreate_productVariant {
  __typename: "ProductVariant";
  id: string;
  stocks: (AddOrRemoveStocks_productVariantStocksCreate_productVariant_stocks | null)[] | null;
}

export interface AddOrRemoveStocks_productVariantStocksCreate {
  __typename: "ProductVariantStocksCreate";
  errors: AddOrRemoveStocks_productVariantStocksCreate_errors[];
  productVariant: AddOrRemoveStocks_productVariantStocksCreate_productVariant | null;
}

export interface AddOrRemoveStocks_productVariantStocksDelete_errors {
  __typename: "StockError";
  code: StockErrorCode;
  field: string | null;
}

export interface AddOrRemoveStocks_productVariantStocksDelete_productVariant_stocks_warehouse {
  __typename: "Warehouse";
  id: string;
  name: string;
}

export interface AddOrRemoveStocks_productVariantStocksDelete_productVariant_stocks {
  __typename: "Stock";
  id: string;
  quantity: number;
  quantityAllocated: number;
  warehouse: AddOrRemoveStocks_productVariantStocksDelete_productVariant_stocks_warehouse;
}

export interface AddOrRemoveStocks_productVariantStocksDelete_productVariant {
  __typename: "ProductVariant";
  id: string;
  stocks: (AddOrRemoveStocks_productVariantStocksDelete_productVariant_stocks | null)[] | null;
}

export interface AddOrRemoveStocks_productVariantStocksDelete {
  __typename: "ProductVariantStocksDelete";
  errors: AddOrRemoveStocks_productVariantStocksDelete_errors[];
  productVariant: AddOrRemoveStocks_productVariantStocksDelete_productVariant | null;
}

export interface AddOrRemoveStocks {
  productVariantStocksCreate: AddOrRemoveStocks_productVariantStocksCreate | null;
  productVariantStocksDelete: AddOrRemoveStocks_productVariantStocksDelete | null;
}

export interface AddOrRemoveStocksVariables {
  variantId: string;
  add: StockInput[];
  remove: string[];
}
