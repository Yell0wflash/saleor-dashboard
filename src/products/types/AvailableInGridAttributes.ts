/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: AvailableInGridAttributes
// ====================================================

export interface AvailableInGridAttributes_attributes_edges_node {
  __typename: "Attribute";
  id: string;
  name: string | null;
}

export interface AvailableInGridAttributes_attributes_edges {
  __typename: "AttributeCountableEdge";
  node: AvailableInGridAttributes_attributes_edges_node;
}

export interface AvailableInGridAttributes_attributes_pageInfo {
  __typename: "PageInfo";
  endCursor: string | null;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  startCursor: string | null;
}

export interface AvailableInGridAttributes_attributes {
  __typename: "AttributeCountableConnection";
  edges: AvailableInGridAttributes_attributes_edges[];
  pageInfo: AvailableInGridAttributes_attributes_pageInfo;
  totalCount: number | null;
}

export interface AvailableInGridAttributes {
  attributes: AvailableInGridAttributes_attributes | null;
}

export interface AvailableInGridAttributesVariables {
  first: number;
  after?: string | null;
}
