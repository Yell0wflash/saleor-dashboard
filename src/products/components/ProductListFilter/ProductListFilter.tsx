import React from "react";
import { useIntl } from "react-intl";

import { FieldType, IFilter } from "@saleor/components/Filter";
import FilterBar from "@saleor/components/FilterBar";
import { FilterProps } from "@saleor/types";
import { StockAvailability } from "@saleor/types/globalTypes";

type ProductListFilterProps = Omit<
  FilterProps,
  "allTabLabel" | "filterLabel" | "searchPlaceholder"
>;

export enum ProductFilterKeys {
  published = "published",
  price = "price",
  priceEqual = "priceEqual",
  priceRange = "priceRange",
  stock = "stock"
}

const ProductListFilter: React.FC<ProductListFilterProps> = props => {
  const intl = useIntl();

  const filterMenu: IFilter<ProductFilterKeys> = [
    {
      children: [],
      data: {
        additionalText: intl.formatMessage({
          defaultMessage: "is set as",
          description: "product status is set as"
        }),
        fieldLabel: intl.formatMessage({
          defaultMessage: "Status",
          description: "product status"
        }),
        options: [
          {
            label: intl.formatMessage({
              defaultMessage: "Visible",
              description: "product is visible"
            }),
            value: true
          },
          {
            label: intl.formatMessage({
              defaultMessage: "Hidden",
              description: "product is hidden"
            }),
            value: false
          }
        ],
        type: FieldType.select
      },
      label: intl.formatMessage({
        defaultMessage: "Visibility",
        description: "product visibility"
      }),
      value: ProductFilterKeys.published
    },
    {
      children: [],
      data: {
        fieldLabel: intl.formatMessage({
          defaultMessage: "Stock quantity"
        }),
        options: [
          {
            label: intl.formatMessage({
              defaultMessage: "Available",
              description: "product status"
            }),
            value: StockAvailability.IN_STOCK
          },
          {
            label: intl.formatMessage({
              defaultMessage: "Out Of Stock",
              description: "product status"
            }),
            value: StockAvailability.OUT_OF_STOCK
          }
        ],
        type: FieldType.select
      },
      label: intl.formatMessage({
        defaultMessage: "Stock",
        description: "product stock"
      }),
      value: ProductFilterKeys.stock
    },
    {
      children: [
        {
          children: [],
          data: {
            additionalText: intl.formatMessage({
              defaultMessage: "equals",
              description: "product price"
            }),
            fieldLabel: null,
            type: FieldType.price
          },
          label: intl.formatMessage({
            defaultMessage: "Specific Price"
          }),
          value: ProductFilterKeys.priceEqual
        },
        {
          children: [],
          data: {
            fieldLabel: intl.formatMessage({
              defaultMessage: "Range"
            }),
            type: FieldType.rangePrice
          },
          label: intl.formatMessage({
            defaultMessage: "Range"
          }),
          value: ProductFilterKeys.priceRange
        }
      ],
      data: {
        fieldLabel: intl.formatMessage({
          defaultMessage: "Price"
        }),
        type: FieldType.range
      },
      label: intl.formatMessage({
        defaultMessage: "Price"
      }),
      value: ProductFilterKeys.price
    }
  ];

  return (
    <FilterBar
      {...props}
      allTabLabel={intl.formatMessage({
        defaultMessage: "All Products",
        description: "tab name"
      })}
      filterMenu={filterMenu}
      filterLabel={intl.formatMessage({
        defaultMessage: "Select all products where:"
      })}
      searchPlaceholder={intl.formatMessage({
        defaultMessage: "Search Products..."
      })}
    />
  );
};
ProductListFilter.displayName = "ProductListFilter";
export default ProductListFilter;
