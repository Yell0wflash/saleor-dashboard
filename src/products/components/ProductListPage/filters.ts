import { defineMessages, IntlShape } from "react-intl";

import { FilterOpts, MinMax } from "@saleor/types";
import { StockAvailability } from "@saleor/types/globalTypes";
import {
  createOptionsField,
  createPriceField
} from "@saleor/utils/filters/fields";
import { IFilter } from "@saleor/components/Filter";

export enum ProductFilterKeys {
  status = "status",
  price = "price",
  stock = "stock"
}

export interface ProductListFilterOpts {
  price: FilterOpts<MinMax>;
  status: FilterOpts<ProductStatus>;
  stockStatus: FilterOpts<StockAvailability>;
}

export enum ProductStatus {
  PUBLISHED = "published",
  HIDDEN = "hidden"
}

const messages = defineMessages({
  available: {
    defaultMessage: "Available",
    description: "product status"
  },
  hidden: {
    defaultMessage: "Hidden",
    description: "product is hidden"
  },
  outOfStock: {
    defaultMessage: "Out Of Stock",
    description: "product status"
  },
  price: {
    defaultMessage: "Price"
  },
  quantity: {
    defaultMessage: "Stock quantity",
    description: "product"
  },
  visibility: {
    defaultMessage: "Visibility",
    description: "product visibility"
  },
  visible: {
    defaultMessage: "Visible",
    description: "product is visible"
  }
});

export function createFilterStructure(
  intl: IntlShape,
  opts: ProductListFilterOpts
): IFilter<ProductFilterKeys> {
  return [
    {
      ...createOptionsField(
        ProductFilterKeys.status,
        intl.formatMessage(messages.visibility),
        [opts.status.value],
        false,
        [
          {
            label: intl.formatMessage(messages.visible),
            value: ProductStatus.PUBLISHED
          },
          {
            label: intl.formatMessage(messages.hidden),
            value: ProductStatus.HIDDEN
          }
        ]
      ),
      active: opts.status.active
    },
    {
      ...createOptionsField(
        ProductFilterKeys.stock,
        intl.formatMessage(messages.quantity),
        [opts.stockStatus.value],
        false,
        [
          {
            label: intl.formatMessage(messages.available),
            value: StockAvailability.IN_STOCK
          },
          {
            label: intl.formatMessage(messages.outOfStock),
            value: StockAvailability.OUT_OF_STOCK
          }
        ]
      ),
      active: opts.stockStatus.active
    },
    {
      ...createPriceField(
        ProductFilterKeys.price,
        intl.formatMessage(messages.price),
        opts.price.value
      ),
      active: opts.price.active
    }
  ];
}
