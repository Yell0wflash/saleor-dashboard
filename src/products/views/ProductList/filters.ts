import { defineMessages, IntlShape } from "react-intl";

import { FilterContentSubmitData } from "../../../components/Filter";
import { Filter } from "../../../components/TableFilter";
import {
  ProductFilterInput,
  StockAvailability
} from "../../../types/globalTypes";
import {
  createFilterTabUtils,
  createFilterUtils
} from "../../../utils/filters";
import { ProductFilterKeys } from "../../components/ProductListFilter";
import {
  ProductListUrlFilters,
  ProductListUrlFiltersEnum,
  ProductListUrlQueryParams
} from "../../urls";

export const PRODUCT_FILTERS_KEY = "productFilters";

export function getFilterVariables(
  params: ProductListUrlFilters
): ProductFilterInput {
  return {
    isPublished:
      params.isPublished !== undefined ? params.isPublished === "true" : null,
    price: {
      gte: parseFloat(params.priceFrom),
      lte: parseFloat(params.priceTo)
    },
    search: params.query,
    stockAvailability: StockAvailability[params.status]
  };
}

export function createFilter(
  filter: FilterContentSubmitData<ProductFilterKeys>
): ProductListUrlFilters {
  const filterName = filter.name;
  if (filterName === ProductFilterKeys.priceEqual) {
    const value = filter.value as string;
    return {
      priceFrom: value,
      priceTo: value
    };
  } else if (filterName === ProductFilterKeys.priceRange) {
    const { value } = filter;
    return {
      priceFrom: value[0],
      priceTo: value[1]
    };
  } else if (filterName === ProductFilterKeys.published) {
    return {
      isPublished: filter.value as string
    };
  } else if (filterName === ProductFilterKeys.stock) {
    const value = filter.value as string;
    return {
      status: StockAvailability[value]
    };
  }
}

const filterMessages = defineMessages({
  available: {
    defaultMessage: "Available",
    description: "filter products by stock"
  },
  hidden: {
    defaultMessage: "Hidden",
    description: "filter products by visibility"
  },
  outOfStock: {
    defaultMessage: "Out of stock",
    description: "filter products by stock"
  },
  priceFrom: {
    defaultMessage: "Price from {price}",
    description: "filter by price"
  },
  priceIs: {
    defaultMessage: "Price is {price}",
    description: "filter by price"
  },
  priceTo: {
    defaultMessage: "Price to {price}",
    description: "filter by price"
  },
  published: {
    defaultMessage: "Published",
    description: "filter products by visibility"
  }
});

interface ProductListChipFormatData {
  currencySymbol: string;
  locale: string;
}
export function createFilterChips(
  filters: ProductListUrlFilters,
  formatData: ProductListChipFormatData,
  onFilterDelete: (filters: ProductListUrlFilters) => void,
  intl: IntlShape
): Filter[] {
  let filterChips: Filter[] = [];

  if (!!filters.priceFrom || !!filters.priceTo) {
    if (filters.priceFrom === filters.priceTo) {
      filterChips = [
        ...filterChips,
        {
          label: intl.formatMessage(filterMessages.priceIs, {
            price: parseFloat(filters.priceFrom).toLocaleString(
              formatData.locale,
              {
                currency: formatData.currencySymbol,
                style: "currency"
              }
            )
          }),
          onClick: () =>
            onFilterDelete({
              ...filters,
              priceFrom: undefined,
              priceTo: undefined
            })
        }
      ];
    } else {
      if (!!filters.priceFrom) {
        filterChips = [
          ...filterChips,
          {
            label: intl.formatMessage(filterMessages.priceFrom, {
              price: parseFloat(filters.priceFrom).toLocaleString(
                formatData.locale,
                {
                  currency: formatData.currencySymbol,
                  style: "currency"
                }
              )
            }),
            onClick: () =>
              onFilterDelete({
                ...filters,
                priceFrom: undefined
              })
          }
        ];
      }

      if (!!filters.priceTo) {
        filterChips = [
          ...filterChips,
          {
            label: intl.formatMessage(filterMessages.priceTo, {
              price: parseFloat(filters.priceTo).toLocaleString(
                formatData.locale,
                {
                  currency: formatData.currencySymbol,
                  style: "currency"
                }
              )
            }),
            onClick: () =>
              onFilterDelete({
                ...filters,
                priceTo: undefined
              })
          }
        ];
      }
    }
  }

  if (!!filters.status) {
    filterChips = [
      ...filterChips,
      {
        label:
          filters.status === StockAvailability.IN_STOCK.toString()
            ? intl.formatMessage(filterMessages.available)
            : intl.formatMessage(filterMessages.outOfStock),
        onClick: () =>
          onFilterDelete({
            ...filters,
            status: undefined
          })
      }
    ];
  }

  if (!!filters.isPublished) {
    filterChips = [
      ...filterChips,
      {
        label: !!filters.isPublished
          ? intl.formatMessage(filterMessages.published)
          : intl.formatMessage(filterMessages.hidden),
        onClick: () =>
          onFilterDelete({
            ...filters,
            isPublished: undefined
          })
      }
    ];
  }

  return filterChips;
}

export const {
  deleteFilterTab,
  getFilterTabs,
  saveFilterTab
} = createFilterTabUtils<ProductListUrlFilters>(PRODUCT_FILTERS_KEY);

export const { areFiltersApplied, getActiveFilters } = createFilterUtils<
  ProductListUrlQueryParams,
  ProductListUrlFilters
>(ProductListUrlFiltersEnum);
