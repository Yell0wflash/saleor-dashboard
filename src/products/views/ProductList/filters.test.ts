import { createIntl } from "react-intl";

import { ProductFilterKeys } from "@saleor/products/components/ProductListFilter";
import { StockAvailability } from "@saleor/types/globalTypes";
import { createFilter, createFilterChips, getFilterVariables } from "./filters";

const mockIntl = createIntl({
  locale: "en"
});

describe("Create filter object", () => {
  it("with price", () => {
    const filter = createFilter({
      name: ProductFilterKeys.priceEqual,
      value: "10"
    });

    expect(filter).toMatchSnapshot();
  });

  it("with price range", () => {
    const filter = createFilter({
      name: ProductFilterKeys.priceEqual,
      value: ["10", "20"]
    });

    expect(filter).toMatchSnapshot();
  });

  it("with publication status", () => {
    const filter = createFilter({
      name: ProductFilterKeys.published,
      value: "false"
    });

    expect(filter).toMatchSnapshot();
  });

  it("with stock status", () => {
    const filter = createFilter({
      name: ProductFilterKeys.stock,
      value: StockAvailability.OUT_OF_STOCK
    });

    expect(filter).toMatchSnapshot();
  });
});

test("Crate filter chips", () => {
  const chips = createFilterChips(
    {
      isPublished: "true",
      priceFrom: "10",
      priceTo: "20",
      status: StockAvailability.IN_STOCK
    },
    {
      currencySymbol: "USD",
      locale: "en"
    },
    jest.fn(),
    mockIntl as any
  );

  expect(chips).toMatchSnapshot();
});

test("Get filter variables", () => {
  const filter = getFilterVariables({
    isPublished: "true",
    priceFrom: "10",
    priceTo: "20",
    status: StockAvailability.IN_STOCK
  });

  expect(filter).toMatchSnapshot();
});
