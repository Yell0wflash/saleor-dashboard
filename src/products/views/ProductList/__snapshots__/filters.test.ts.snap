// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crate filter chips 1`] = `
Array [
  Object {
    "label": "Price from $10.00",
    "onClick": [Function],
  },
  Object {
    "label": "Price to $20.00",
    "onClick": [Function],
  },
  Object {
    "label": "Available",
    "onClick": [Function],
  },
  Object {
    "label": "Published",
    "onClick": [Function],
  },
]
`;

exports[`Create filter object with price 1`] = `
Object {
  "priceFrom": "10",
  "priceTo": "10",
}
`;

exports[`Create filter object with price range 1`] = `
Object {
  "priceFrom": Array [
    "10",
    "20",
  ],
  "priceTo": Array [
    "10",
    "20",
  ],
}
`;

exports[`Create filter object with publication status 1`] = `
Object {
  "isPublished": "false",
}
`;

exports[`Create filter object with stock status 1`] = `
Object {
  "status": "OUT_OF_STOCK",
}
`;

exports[`Get filter variables 1`] = `
Object {
  "isPublished": true,
  "price": Object {
    "gte": 10,
    "lte": 20,
  },
  "search": undefined,
  "stockAvailability": "IN_STOCK",
}
`;
